src/main/java/dao/DaoReaderImplementation.java

```diff
@@ -1,33 +1,100 @@
+                }
+            }
+            return Optional.empty();
+        } catch (SQLException e) {
+            throw new DAOException("Database error during retrieval reader id!" + "\nError details: " + e.getMessage());
+        }
+    }
+
+    private Reader mapToReader (ResultSet rs) {
+        try {
+            return new Reader(rs.getInt("id"), rs.getString("name"));
+        } catch (SQLException e) {
+            throw new DAOException("Failed to map resultSet to Reader object!" + "\nError details: " + e.getMessage());
+        }
     }
 
     @Override
-    public Optional<Reader> findById(int id) {
-        return readers.stream().filter(reader -> reader.getId() == id).findFirst();
+    public Reader save (Reader readerToSave) {
+        try (var connection = ConnectionUtil.createConnection();
+             var statement = connection.prepareStatement("insert into reader(name) values(?)", Statement.RETURN_GENERATED_KEYS)) {
+            statement.setString(1, readerToSave.getName());
+            statement.executeUpdate();
+            try(var generatedKeys =  statement.getGeneratedKeys()) {
+                if (generatedKeys.next()) {
+                    readerToSave.setId(generatedKeys.getInt(1));
+                    return readerToSave;
+                } else {
+                    throw new DAOException("Creating reader failed, no ID obtained!");
+                }
+            }
+        } catch (SQLException e) {
+            throw new DAOException("Database error during saving new reader!" + "\nError details: " + e.getMessage());
+        }
     }
 
     @Override
-    public Reader save(Reader readerToSave) {
-        readers.add(readerToSave);
-        return readerToSave;
+    public Optional<Reader> findByBookId(long bookId) {
+        try (var connection = ConnectionUtil.createConnection();
+             var statement = connection.prepareStatement("select reader.id, reader.name from reader inner join book on reader.id = book.readerid where book.id = ?")) {
+            statement.setLong(1, bookId);
+            try (var resultSet = statement.executeQuery()) {
+                while (resultSet.next()) {
+                    return Optional.of(mapToReader(resultSet));
+                }
+            }
+            return Optional.empty();
+        } catch (SQLException e) {
+            throw new DAOException("Database error during retrieval reader by book Id!" + "\nError details: " + e.getMessage());
+        }
+    }
+    public Map<Reader, Optional<Book>> findAllWithBooks () {
+        try (var connection = ConnectionUtil.createConnection();
+             var statement = connection.prepareStatement("select reader.id, reader.name, book.id from reader inner join book on reader.id = book.readerid order by reader.id")) {
+            Map<Reader, Optional<Book>> map = new HashMap<>();
+            DaoBookImplementation daoBookImplementation = new DaoBookImplementation();
```

Comment:
Когда SQL запросы становятся длиннее нескольких слов - есть смысл форматировать их и выносить в отдельную переменную. Так проще их читать и редактировать.

```suggestion
    public Map<Reader, Optional<Book>> findAllWithBooks () {
       var query = """
           select 
               reader.id, 
               reader.name, 
               book.id 
            from reader 
               inner join book on reader.id = book.readerid 
            order by reader.id
         """;
        try (var connection = ConnectionUtil.createConnection();
             var statement = connection.prepareStatement(query)) {
            Map<Reader, Optional<Book>> map = new HashMap<>();
```